libblkid/src/superblocks/zfs.c

```diff
@@ -255,72 +287,68 @@ static int find_uberblocks(const void *label, loff_t *ub_offset, int *swap_endia
-	loff_t offset = 0, ub_offset = 0;
-	int label_no, found = 0, found_in_label;
-	const void *label;
-	loff_t blk_align = (pr->size % (256 * 1024ULL));
+	loff_t offset = 0;
+	int label_no;
+	struct nvs_header_t *label = NULL;
+	bool found_label = false;
 
 	DBG(PROBE, ul_debug("probe_zfs"));
-	/* Look for at least 4 uberblocks to ensure a positive match */
+
+	if (pr->size < ZFS_MINDEVSIZE)
+		return (1);
+
+	/* Look for at least one valid label to ensure a positive match */
 	for (label_no = 0; label_no < 4; label_no++) {
-		switch(label_no) {
-		case 0: // jump to L0
-			offset = 0;
-			break;
-		case 1: // jump to L1
-			offset = VDEV_LABEL_SIZE;
-			break;
-		case 2: // jump to L2
-			offset = pr->size - 2 * VDEV_LABEL_SIZE - blk_align;
-			break;
-		case 3: // jump to L3
-			offset = pr->size - VDEV_LABEL_SIZE - blk_align;
-			break;
-		}
+		offset = label_offset(pr->size, label_no) + VDEV_LABEL_NVPAIR;
 
 		if ((S_ISREG(pr->mode) || blkid_probe_is_wholedisk(pr)) &&
-		    blkid_probe_is_covered_by_pt(pr,  offset, VDEV_LABEL_SIZE))
+		    blkid_probe_is_covered_by_pt(pr,  offset, VDEV_PHYS_SIZE))
 			/* ignore this area, it's within any partition and
 			 * we are working with whole-disk now */
 			continue;
 
-		label = blkid_probe_get_buffer(pr, offset, VDEV_LABEL_SIZE);
-		if (label == NULL)
-			return errno ? -errno : 1;
+		label = (struct nvs_header_t *) blkid_probe_get_buffer(pr, offset, VDEV_PHYS_SIZE);
 
-		found_in_label = find_uberblocks(label, &ub_offset, &swab_endian);
+		/*
+		 * Label supports XDR encoding, reject for any other unsupported format. Also
+		 * endianess can be 0 or 1, reject garbage value. Moreover, check if first
+		 * nvpair encode size is non-zero.
+		 */
+		if (!label || label->nvh_encoding != 0x1 || !be32_to_cpu(label->nvh_first_size) ||
+		    (unsigned char) label->nvh_endian > 0x1)
+			continue;
 
-		if (found_in_label > 0) {
-			found+= found_in_label;
-			ub = (struct zfs_uberblock *)((char *) label + ub_offset);
-			ub_offset += offset;
+		if (host_endian != label->nvh_endian)
+			swab_endian = 1;
 
-			if (found >= ZFS_WANT)
-				break;
+		if (zfs_extract_guid_name(pr, label, VDEV_PHYS_SIZE, true)) {
+			found_label = true;
+			break;
 		}
 	}
 
-	if (found < ZFS_WANT)
-		return 1;
-
-	/* If we found the 4th uberblock, then we will have exited from the
-	 * scanning loop immediately, and ub will be a valid uberblock. */
-	blkid_probe_sprintf_version(pr, "%" PRIu64, swab_endian ?
-				    swab64(ub->ub_version) : ub->ub_version);
+	if (!label || !found_label)
+		return (1);
 
-	zfs_extract_guid_name(pr, offset);
+	(void) zfs_extract_guid_name(pr, label, VDEV_PHYS_SIZE, false);
 
-	if (blkid_probe_set_magic(pr, ub_offset,
-				sizeof(ub->ub_magic),
-				(unsigned char *) &ub->ub_magic))
-		return 1;
+	/*
+	 * Zero out whole nvlist header including fisrt nvpair size
+	 */
+	if (blkid_probe_set_magic(pr, offset, sizeof(struct nvs_header_t),
+	    (unsigned char *) &label->nvh_first_size))
+		return (1);
 
```

Comment:
@ixhamza, Coverity scan reported this as
```
>>>     CID 451238:  Memory - corruptions  (OVERRUN)
>>>     Overrunning buffer pointed to by "(unsigned char *)&label->nvh_first_size" of 4 bytes by passing it to a function which accesses it at byte offset 15 using argument "16UL".
```
This is a false positive, as the label is large enough, but the code seems unintuitive.          
                                                                                                 
It uses sizeof(struct nvs_header_t) in conjunction with a pointer to the last                    
item in the struct. This does not seem very readable.  Would it be better to                     
zero out the entire label (struct nvs_header_t)? Since label->nvh_first_size                     
cannot be zero, this would mean:                                                                 
                                                                                                 
        blkid_probe_set_magic(pr, offset, sizeof(struct nvs_header_t), (unsigned char *) label)) 
                                                                                                 
Alternatively, we could use sizeof(struct nvlist) to zero out the label and the first pair, which
would mean:
                                                                                                 
        blkid_probe_set_magic(pr, offset, sizeof(struct nvs_nvlist), (unsigned char *) label))

Reply:
@karelzak - `blkid_probe_set_magic(pr, offset, sizeof(struct nvs_header_t), (unsigned char *) label))` sounds much reasonable to me. I can open a PR after testing!

Reply:
https://github.com/util-linux/util-linux/pull/3295